src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -142,4 +142,78 @@ public void close() throws IOException {
         ch.close();
     }
 
+    private static final int TRANSFER_SIZE = 8192;
+
+    @Override
+    public long transferTo(OutputStream out) throws IOException {
+        if (out instanceof ChannelOutputStream cos) {
+            WritableByteChannel oc = cos.channel();
+            long i = 0L;
+
+            if (ch instanceof FileChannel fc) {
+                long pos = fc.position();
+                long size = fc.size();
+                try {
+                    for (long n = size - pos; i < n;)
+                        i += fc.transferTo(pos + i, Long.MAX_VALUE, oc);
+                    return i;
+                } finally {
+                    fc.position(pos + i);
+                }
+            }
+
+            if (oc instanceof FileChannel fc) {
+                long fcpos = fc.position();
+
+                if (ch instanceof SeekableByteChannel sbc) {
+                    long pos = sbc.position();
+                    long size = sbc.size();
+                    try {
+                        for (long n = size - pos; i < n;)
+                            i += fc.transferFrom(ch, fcpos + i, Long.MAX_VALUE);
+                        return i;
+                    } finally {
+                        fc.position(fcpos + i);
+                    }
+                }
+
+                ByteBuffer bb = Util.getTemporaryDirectBuffer(TRANSFER_SIZE);
+                try {
+                    int r;
+                    do {
+                        i += fc.transferFrom(ch, fcpos + i, Long.MAX_VALUE);
+                        r = ch.read(bb); // detect end-of-stream
+                        if (r > -1) {
+                            bb.flip();
+                            while (bb.hasRemaining())
+                                oc.write(bb);
+                            bb.clear();
+                            i += r;
+                        }
+                    } while (r > -1);
+                    return i;
+                } finally {
+                    fc.position(fcpos + i);
+                    Util.releaseTemporaryDirectBuffer(bb);
+                }
+            }
+
+            ByteBuffer bb = Util.getTemporaryDirectBuffer(TRANSFER_SIZE);
+            try {
+                for (int r = ch.read(bb); r > -1; r = ch.read(bb)) {
+                    bb.flip();
+                    while (bb.hasRemaining())
+                        oc.write(bb);
+                    bb.clear();
+                    i += r;
+                }
+                return i;
+            } finally {
+                Util.releaseTemporaryDirectBuffer(bb);
+            }
+        }
+
+        return super.transferTo(out);
```

Comment:
(i am no reviewer)
wouldn't this method be more intuitive if it wouldn't try to avoid using "else if" and "else"? If there are sequential if blocks with return in them, part of me is always trying to find the fall-through scenario, but in this case its all distinct code paths. Using branches would make that obvious + if there would be a fall through in future, the compiler would generate a "missing return" error right away. 

example:
```java
        if (out instanceof ChannelOutputStream cos) {
            
            WritableByteChannel oc = cos.channel();
            long i = 0L;

            if (ch instanceof FileChannel fc) {
                
                long pos = fc.position();
                long size = fc.size();
                try {
                    for (long n = size - pos; i < n;)
                        i += fc.transferTo(pos + i, Long.MAX_VALUE, oc);
                    return i;
                } finally {
                    fc.position(pos + i);
                }
                
            } else if (oc instanceof FileChannel fc) {
                
                long fcpos = fc.position();

                if (ch instanceof SeekableByteChannel sbc) {
                    
                    long pos = sbc.position();
                    long size = sbc.size();
                    try {
                        for (long n = size - pos; i < n;)
                            i += fc.transferFrom(ch, fcpos + i, Long.MAX_VALUE);
                        return i;
                    } finally {
                        fc.position(fcpos + i);
                    }
                    
                } else {
                    
                    ByteBuffer bb = Util.getTemporaryDirectBuffer(TRANSFER_SIZE);
                    try {
                        int r;
                        do {
                            i += fc.transferFrom(ch, fcpos + i, Long.MAX_VALUE);
                            r = ch.read(bb); // detect end-of-stream
                            if (r > -1) {
                                bb.flip();
                                while (bb.hasRemaining())
                                    oc.write(bb);
                                bb.clear();
                                i += r;
                            }
                        } while (r > -1);
                        return i;
                    } finally {
                        fc.position(fcpos + i);
                        Util.releaseTemporaryDirectBuffer(bb);
                    }
                }

            } else {
                
                ByteBuffer bb = Util.getTemporaryDirectBuffer(TRANSFER_SIZE);
                try {
                    for (int r = ch.read(bb); r > -1; r = ch.read(bb)) {
                        bb.flip();
                        while (bb.hasRemaining())
                            oc.write(bb);
                        bb.clear();
                        i += r;
                    }
                    return i;
                } finally {
                    Util.releaseTemporaryDirectBuffer(bb);
                }
            }
        } else {
            return super.transferTo(out);
        }
```

Reply:
This is just about personal likes and dislikes. In fact, there is no "missing" case at all as this is not about "there must be THIS solution OR that solution", but this is just about doing "the standard case" vs. doing optimizations "if applicable" -- which is a different semantic. I personally dislike your proposal as it makes the code overly cluttered and hard to read. No other reviewer wanted it that way. So until a majority of reviewers actually wants me to clutter the code with else cases, I'd rather abstain from your proposal.